File: lawclick-next/src/lib/permissions.ts
/**
 * 权限配置 - 基于架构设计文档的身份权限管理
 * 
 * 角色层级：
 * - 专业序列：PARTNER > SENIOR_LAWYER > LAWYER > TRAINEE
 * - 行政序列：ADMIN > HR/MARKETING > LEGAL_SECRETARY
 * - 外部角色：CLIENT
 * - 系统角色：FIRM_ENTITY（用于离职归档）
 */

import type { Role } from "@prisma/client"
import type { RoleKey } from "@/lib/role-keys"

type Assert<T extends true> = T
type RoleKeysMatchPrismaRole = Exclude<Role, RoleKey> extends never ? (Exclude<RoleKey, Role> extends never ? true : false) : false
export const ROLE_KEYS_MATCH_PRISMA_ROLE: Assert<RoleKeysMatchPrismaRole> = true

// ==============================================================================
// 权限类型定义
// ==============================================================================

export type Permission =
    // Dashboard permissions
    | 'dashboard:view'
    | 'dashboard:edit'
    // Case permissions
    | 'case:view'
    | 'case:create'
    | 'case:edit'
    | 'case:delete'
    | 'case:assign'
    | 'case:archive'
    // Task permissions
    | 'task:view'
    | 'task:create'
    | 'task:edit'
    | 'task:delete'
    // Document permissions
    | 'document:view'
    | 'document:upload'
    | 'document:edit'
    | 'document:delete'
    | 'document:template_manage'
    // Billing permissions
    | 'billing:view'
    | 'billing:create'
    | 'billing:edit'
    | 'billing:approve'
    | 'timelog:approve'
    // Team permissions
    | 'team:view'
    | 'team:manage'
    | 'user:manage'
    | 'user:view_all'
    // Approval / OA permissions
    | 'approval:create'
    | 'approval:approve'
    | 'approval:view_all'
    // CRM permissions
    | 'crm:view'
    | 'crm:edit'
  // Tools permissions
  | 'tools:manage'
  // AI permissions
  | 'ai:use'
  // Admin permissions
  | 'admin:access'
  | 'admin:settings'
  | 'admin:audit'

export const ALL_PERMISSIONS = [
    "dashboard:view",
    "dashboard:edit",
    "case:view",
    "case:create",
    "case:edit",
    "case:delete",
    "case:assign",
    "case:archive",
    "task:view",
    "task:create",
    "task:edit",
    "task:delete",
    "document:view",
    "document:upload",
    "document:edit",
    "document:delete",
    "document:template_manage",
    "billing:view",
    "billing:create",
    "billing:edit",
    "billing:approve",
    "timelog:approve",
    "team:view",
    "team:manage",
    "user:manage",
    "user:view_all",
    "approval:create",
    "approval:approve",
    "approval:view_all",
    "crm:view",
    "crm:edit",
    "tools:manage",
    "ai:use",
    "admin:access",
    "admin:settings",
    "admin:audit",
] as const satisfies ReadonlyArray<Permission>

type PermissionKeysMatch = Exclude<Permission, (typeof ALL_PERMISSIONS)[number]> extends never
    ? Exclude<(typeof ALL_PERMISSIONS)[number], Permission> extends never
        ? true
        : false
    : false
export const PERMISSION_KEYS_MATCH_ALL_PERMISSIONS: Assert<PermissionKeysMatch> = true

// ==============================================================================
// 角色层级（数字越大权限越高）
// ==============================================================================

export const ROLE_HIERARCHY: Record<Role, number> = {
    PARTNER: 100,
    SENIOR_LAWYER: 80,
    LAWYER: 60,
    TRAINEE: 40,
    ADMIN: 90,
    HR: 50,
    MARKETING: 50,
    LEGAL_SECRETARY: 45,
    CLIENT: 10,
    FIRM_ENTITY: 0,
}

// ==============================================================================
// 角色权限配置
// ==============================================================================

type RolePermissions = Record<Role, Permission[]>

export const ROLE_PERMISSIONS: RolePermissions = {
    PARTNER: [
        'dashboard:view', 'dashboard:edit',
        'case:view', 'case:create', 'case:edit', 'case:delete', 'case:assign', 'case:archive',
        'task:view', 'task:create', 'task:edit', 'task:delete',
        'document:view', 'document:upload', 'document:edit', 'document:delete', 'document:template_manage',
        'billing:view', 'billing:create', 'billing:edit', 'billing:approve', 'timelog:approve',
        'team:view', 'team:manage', 'user:manage', 'user:view_all',
        'approval:create', 'approval:approve', 'approval:view_all',
        'crm:view', 'crm:edit',
        'tools:manage',
        'ai:use',
        'admin:access', 'admin:settings', 'admin:audit'
    ],
    SENIOR_LAWYER: [
        'dashboard:view', 'dashboard:edit',
        'case:view', 'case:create', 'case:edit', 'case:assign', 'case:archive',
        'task:view', 'task:create', 'task:edit', 'task:delete',
        'document:view', 'document:upload', 'document:edit', 'document:delete',
        'billing:view', 'billing:create', 'timelog:approve',
        'team:view', 'team:manage',
        'approval:create', 'approval:approve',
        'crm:view', 'crm:edit',
        'ai:use',
    ],
    LAWYER: [
        'dashboard:view', 'dashboard:edit',
        'case:view', 'case:create', 'case:edit', 'case:assign',
        'task:view', 'task:create', 'task:edit',
        'document:view', 'document:upload', 'document:edit',
        'billing:view', 'billing:create',
        'team:view',
        'approval:create',
        'crm:view', 'crm:edit',
        'ai:use',
    ],
    TRAINEE: [
        'dashboard:view', 'dashboard:edit',
        'case:view',
        'task:view', 'task:create', 'task:edit',
        'document:view', 'document:upload',
        'team:view',
        'approval:create',
        'crm:view',
        'ai:use',
    ],
    ADMIN: [
        'dashboard:view', 'dashboard:edit',
        'case:view',
        'task:view', 'task:create', 'task:edit', 'task:delete',
        'document:view', 'document:upload', 'document:edit', 'document:delete', 'document:template_manage',
        'billing:view', 'billing:create', 'billing:edit', 'billing:approve',
        'team:view', 'team:manage', 'user:manage', 'user:view_all',
        'approval:create', 'approval:approve', 'approval:view_all',
        'crm:view', 'crm:edit',
        'tools:manage',
        'ai:use',
        'admin:access', 'admin:settings', 'admin:audit'
    ],
    HR: [
        'dashboard:view', 'dashboard:edit',
        'task:view', 'task:create', 'task:edit',
        'document:view', 'document:upload',
        'team:view', 'user:manage', 'user:view_all',
        'approval:create',
        'ai:use',
    ],
    MARKETING: [
        'dashboard:view', 'dashboard:edit',
        'task:view', 'task:create', 'task:edit',
        'document:view', 'document:upload',
        'team:view',
        'approval:create',
        'crm:view', 'crm:edit',
        'ai:use',
    ],
    LEGAL_SECRETARY: [
        'dashboard:view', 'dashboard:edit',
        'case:view',
        'task:view', 'task:create', 'task:edit',
        'document:view', 'document:upload', 'document:edit',
        'billing:view',
        'team:view',
        'approval:create',
        'crm:view', 'crm:edit',
        'ai:use',
    ],
    CLIENT: [
        'case:view',
        'document:view',
        'billing:view'
    ],
    FIRM_ENTITY: []
}

// ==============================================================================
// 辅助函数
// ==============================================================================

export function hasPermission(role: Role, permission: Permission): boolean {
    const permissions = ROLE_PERMISSIONS[role] || []
    return permissions.includes(permission)
}

export function hasAnyPermission(role: Role, permissions: Permission[]): boolean {
    return permissions.some(p => hasPermission(role, p))
}

export function hasAllPermissions(role: Role, permissions: Permission[]): boolean {
    return permissions.every(p => hasPermission(role, p))
}

export function isHigherRole(role1: Role, role2: Role): boolean {
    return ROLE_HIERARCHY[role1] > ROLE_HIERARCHY[role2]
}

// ==============================================================================
// 角色显示名称
// ==============================================================================

export const ROLE_DISPLAY_NAMES: Record<Role, string> = {
    PARTNER: '合伙人',
    SENIOR_LAWYER: '高级律师',
    LAWYER: '专职律师',
    TRAINEE: '实习生',
    ADMIN: '管理员',
    HR: '人事',
    MARKETING: '品牌',
    LEGAL_SECRETARY: '法律秘书',
    CLIENT: '客户',
    FIRM_ENTITY: '律所',
}

export const PROFESSIONAL_ROLES: Role[] = ['PARTNER', 'SENIOR_LAWYER', 'LAWYER', 'TRAINEE']
export const ADMIN_ROLES: Role[] = ['ADMIN', 'HR', 'MARKETING', 'LEGAL_SECRETARY']

export function isProfessionalRole(role: Role): boolean {
    return PROFESSIONAL_ROLES.includes(role)
}

export function isAdminRole(role: Role): boolean {
    return ADMIN_ROLES.includes(role)
}

File: src/security/permissions.rs
//! 权限模型（Rust 侧）
//!
//! 目标：与 `lawclick-next/src/lib/permissions.ts` 的口径保持一致，
//! 避免出现“同库同用户，但 Rust/Next 权限行为不一致”的隐患。

use crate::entity::user::Role;
use crate::error::{AppError, AppResult};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Permission {
    // Dashboard permissions
    DashboardView,
    DashboardEdit,
    // Case permissions
    CaseView,
    CaseCreate,
    CaseEdit,
    CaseDelete,
    CaseAssign,
    CaseArchive,
    // Task permissions
    TaskView,
    TaskCreate,
    TaskEdit,
    TaskDelete,
    // Document permissions
    DocumentView,
    DocumentUpload,
    DocumentEdit,
    DocumentDelete,
    DocumentTemplateManage,
    // Billing permissions
    BillingView,
    BillingCreate,
    BillingEdit,
    BillingApprove,
    TimeLogApprove,
    // Team permissions
    TeamView,
    TeamManage,
    UserManage,
    UserViewAll,
    // Approval permissions
    ApprovalCreate,
    ApprovalApprove,
    ApprovalViewAll,
    // CRM permissions
    CrmView,
    CrmEdit,
    // Tools permissions
    ToolsManage,
    // AI permissions
    AiUse,
    // Admin permissions
    AdminAccess,
    AdminSettings,
    AdminAudit,
}

impl Permission {
    pub fn as_str(self) -> &'static str {
        match self {
            Permission::DashboardView => "dashboard:view",
            Permission::DashboardEdit => "dashboard:edit",
            Permission::CaseView => "case:view",
            Permission::CaseCreate => "case:create",
            Permission::CaseEdit => "case:edit",
            Permission::CaseDelete => "case:delete",
            Permission::CaseAssign => "case:assign",
            Permission::CaseArchive => "case:archive",
            Permission::TaskView => "task:view",
            Permission::TaskCreate => "task:create",
            Permission::TaskEdit => "task:edit",
            Permission::TaskDelete => "task:delete",
            Permission::DocumentView => "document:view",
            Permission::DocumentUpload => "document:upload",
            Permission::DocumentEdit => "document:edit",
            Permission::DocumentDelete => "document:delete",
            Permission::DocumentTemplateManage => "document:template_manage",
            Permission::BillingView => "billing:view",
            Permission::BillingCreate => "billing:create",
            Permission::BillingEdit => "billing:edit",
            Permission::BillingApprove => "billing:approve",
            Permission::TimeLogApprove => "timelog:approve",
            Permission::TeamView => "team:view",
            Permission::TeamManage => "team:manage",
            Permission::UserManage => "user:manage",
            Permission::UserViewAll => "user:view_all",
            Permission::ApprovalCreate => "approval:create",
            Permission::ApprovalApprove => "approval:approve",
            Permission::ApprovalViewAll => "approval:view_all",
            Permission::CrmView => "crm:view",
            Permission::CrmEdit => "crm:edit",
            Permission::ToolsManage => "tools:manage",
            Permission::AiUse => "ai:use",
            Permission::AdminAccess => "admin:access",
            Permission::AdminSettings => "admin:settings",
            Permission::AdminAudit => "admin:audit",
        }
    }
}

impl std::fmt::Display for Permission {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

pub fn parse_role(role: &str) -> Option<Role> {
    match role.trim().to_uppercase().as_str() {
        "PARTNER" => Some(Role::Partner),
        "SENIOR_LAWYER" => Some(Role::SeniorLawyer),
        "LAWYER" => Some(Role::Lawyer),
        "TRAINEE" => Some(Role::Trainee),
        "ADMIN" => Some(Role::Admin),
        "HR" => Some(Role::Hr),
        "MARKETING" => Some(Role::Marketing),
        "LEGAL_SECRETARY" => Some(Role::LegalSecretary),
        "CLIENT" => Some(Role::Client),
        "FIRM_ENTITY" => Some(Role::FirmEntity),
        _ => None,
    }
}

const PARTNER_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::CaseCreate,
    Permission::CaseEdit,
    Permission::CaseDelete,
    Permission::CaseAssign,
    Permission::CaseArchive,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::TaskDelete,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::DocumentEdit,
    Permission::DocumentDelete,
    Permission::DocumentTemplateManage,
    Permission::BillingView,
    Permission::BillingCreate,
    Permission::BillingEdit,
    Permission::BillingApprove,
    Permission::TimeLogApprove,
    Permission::TeamView,
    Permission::TeamManage,
    Permission::UserManage,
    Permission::UserViewAll,
    Permission::ApprovalCreate,
    Permission::ApprovalApprove,
    Permission::ApprovalViewAll,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::ToolsManage,
    Permission::AiUse,
    Permission::AdminAccess,
    Permission::AdminSettings,
    Permission::AdminAudit,
];

const SENIOR_LAWYER_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::CaseCreate,
    Permission::CaseEdit,
    Permission::CaseAssign,
    Permission::CaseArchive,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::TaskDelete,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::DocumentEdit,
    Permission::DocumentDelete,
    Permission::BillingView,
    Permission::BillingCreate,
    Permission::TimeLogApprove,
    Permission::TeamView,
    Permission::TeamManage,
    Permission::ApprovalCreate,
    Permission::ApprovalApprove,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::AiUse,
];

const LAWYER_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::CaseCreate,
    Permission::CaseEdit,
    Permission::CaseAssign,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::DocumentEdit,
    Permission::BillingView,
    Permission::BillingCreate,
    Permission::TeamView,
    Permission::ApprovalCreate,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::AiUse,
];

const TRAINEE_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::TeamView,
    Permission::ApprovalCreate,
    Permission::CrmView,
    Permission::AiUse,
];

const ADMIN_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::TaskDelete,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::DocumentEdit,
    Permission::DocumentDelete,
    Permission::BillingView,
    Permission::BillingCreate,
    Permission::BillingEdit,
    Permission::BillingApprove,
    Permission::TeamView,
    Permission::TeamManage,
    Permission::UserManage,
    Permission::UserViewAll,
    Permission::ApprovalCreate,
    Permission::ApprovalApprove,
    Permission::ApprovalViewAll,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::ToolsManage,
    Permission::AiUse,
    Permission::AdminAccess,
    Permission::AdminSettings,
    Permission::AdminAudit,
];

const HR_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::TeamView,
    Permission::UserManage,
    Permission::UserViewAll,
    Permission::ApprovalCreate,
    Permission::AiUse,
];

const MARKETING_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::TeamView,
    Permission::ApprovalCreate,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::AiUse,
];

const LEGAL_SECRETARY_PERMISSIONS: &[Permission] = &[
    Permission::DashboardView,
    Permission::DashboardEdit,
    Permission::CaseView,
    Permission::TaskView,
    Permission::TaskCreate,
    Permission::TaskEdit,
    Permission::DocumentView,
    Permission::DocumentUpload,
    Permission::DocumentEdit,
    Permission::BillingView,
    Permission::TeamView,
    Permission::ApprovalCreate,
    Permission::CrmView,
    Permission::CrmEdit,
    Permission::AiUse,
];

const CLIENT_PERMISSIONS: &[Permission] = &[
    Permission::CaseView,
    Permission::DocumentView,
    Permission::BillingView,
];

const FIRM_ENTITY_PERMISSIONS: &[Permission] = &[];

pub fn permissions_for_role(role: Role) -> &'static [Permission] {
    match role {
        Role::Partner => PARTNER_PERMISSIONS,
        Role::SeniorLawyer => SENIOR_LAWYER_PERMISSIONS,
        Role::Lawyer => LAWYER_PERMISSIONS,
        Role::Trainee => TRAINEE_PERMISSIONS,
        Role::Admin => ADMIN_PERMISSIONS,
        Role::Hr => HR_PERMISSIONS,
        Role::Marketing => MARKETING_PERMISSIONS,
        Role::LegalSecretary => LEGAL_SECRETARY_PERMISSIONS,
        Role::Client => CLIENT_PERMISSIONS,
        Role::FirmEntity => FIRM_ENTITY_PERMISSIONS,
    }
}

pub fn has_permission(role: Role, permission: Permission) -> bool {
    permissions_for_role(role).contains(&permission)
}

pub fn require_permission(role: Role, permission: Permission) -> AppResult<()> {
    if has_permission(role, permission) {
        return Ok(());
    }
    Err(AppError::Forbidden(format!("缺少权限：{permission}")))
}
